File: walking_robot_sim/src/tests/correctness/test_ik_with_roll.py
#!/usr/bin/env python3
"""Тесты корректности IK с roll — проверка влияния крена на углы суставов."""

import sys
import os
import math
import numpy as np

sys.path.insert(
    0,
    os.path.join(
        os.path.dirname(__file__), "..", "..", "quadropted_controller", "scripts"
    ),
)

from InverseKinematics import InverseKinematics
from ForwardKinematics import ForwardKinematics


def make_default_stance():
    body = [0.3762, 0.0935]
    legs = [0.0, 0.0955, 0.213, 0.213]
    dx = body[0] * 0.5 + 0.02
    dy = body[1] * 0.5 + legs[1]
    return np.array(
        [
            [dx, dx, -dx, -dx],
            [-dy, dy, -dy, dy],
            [0, 0, 0, 0],
        ]
    )


def test_ik_zero_roll_default_stance():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()
    angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0)

    assert len(angles) == 12
    assert abs(angles[0]) < 0.01, f"hip angle should be ~0, got {angles[0]}"
    assert abs(angles[1] - 0.862) < 0.1, (
        f"thigh angle should be ~0.862, got {angles[1]}"
    )
    assert abs(angles[2] - (-1.883)) < 0.1, (
        f"calf angle should be ~-1.883, got {angles[2]}"
    )
    print("  ✅ test_ik_zero_roll_default_stance")


def test_ik_roll_45_degrees_affects_angles():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()

    angles_no_roll = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0)
    angles_roll_45 = ik.inverse_kinematics(stance, 0, 0, 0.25, math.pi / 4, 0, 0)

    different = any(abs(a - b) > 1e-6 for a, b in zip(angles_no_roll, angles_roll_45))
    assert different, "IK с roll=π/4 должен давать другие углы чем roll=0"
    print("  ✅ test_ik_roll_45_degrees_affects_angles")


def test_ik_roll_45_angles_in_valid_range():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()

    angles = ik.inverse_kinematics(stance, 0, 0, 0.25, math.pi / 4, 0, 0)

    for i, a in enumerate(angles):
        assert abs(a) < math.pi + 0.5, f"Angle {i} out of range: {a}"
    print("  ✅ test_ik_roll_45_angles_in_valid_range")


def test_ik_negative_roll_45():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()

    angles_pos = ik.inverse_kinematics(stance, 0, 0, 0.25, math.pi / 4, 0, 0)
    angles_neg = ik.inverse_kinematics(stance, 0, 0, 0.25, -math.pi / 4, 0, 0)

    different = any(abs(a - b) > 1e-6 for a, b in zip(angles_pos, angles_neg))
    assert different, "IK с roll=+π/4 и roll=-π/4 должны давать разные углы"
    print("  ✅ test_ik_negative_roll_45")


def test_ik_left_right_symmetry_zero_roll():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()

    angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0)

    assert abs(angles[0] - angles[3]) < 0.01, (
        f"hip FR vs FL mismatch: {angles[0]} vs {angles[3]}"
    )
    assert abs(angles[1] - angles[4]) < 0.01, (
        f"thigh FR vs FL mismatch: {angles[1]} vs {angles[4]}"
    )
    assert abs(angles[2] - angles[5]) < 0.01, (
        f"calf FR vs FL mismatch: {angles[2]} vs {angles[5]}"
    )
    print("  ✅ test_ik_left_right_symmetry_zero_roll")


def test_fk_ik_roundtrip_zero_roll():
    body = [0.3762, 0.0935]
    legs = [0.0, 0.0955, 0.213, 0.213]

    fk = ForwardKinematics(body, legs)
    ik = InverseKinematics(body, legs)

    original_angles = [0, 0.3, -0.6] * 4
    foot_pos = fk.forward_kinematics_all_legs(original_angles)

    leg_positions = np.zeros((3, 4))
    for leg in range(4):
        for dim in range(3):
            leg_positions[dim, leg] = foot_pos[leg][dim]

    recovered = ik.inverse_kinematics(leg_positions, 0, 0, 0.25, 0, 0, 0)

    for i in range(12):
        assert abs(recovered[i]) < 6.3, f"Angle {i} out of valid range: {recovered[i]}"
    print("  ✅ test_fk_ik_roundtrip_zero_roll")


def test_ik_small_orientation_angles():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()

    angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0.1, -0.05, 0.02)

    for i, a in enumerate(angles):
        assert abs(a) < math.pi + 0.5, f"Angle {i} out of range: {a}"
    print("  ✅ test_ik_small_orientation_angles")


def test_ik_roll_varies_smoothly():
    ik = InverseKinematics([0.3762, 0.0935], [0.0, 0.0955, 0.213, 0.213])
    stance = make_default_stance()

    angles_0 = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0)
    angles_01 = ik.inverse_kinematics(stance, 0, 0, 0.25, 0.1, 0, 0)
    angles_02 = ik.inverse_kinematics(stance, 0, 0, 0.25, 0.2, 0, 0)

    diff_0_01 = sum(abs(a - b) for a, b in zip(angles_0, angles_01))
    diff_01_02 = sum(abs(a - b) for a, b in zip(angles_01, angles_02))

    assert diff_0_01 > 0, "Углы должны меняться при roll=0.1"
    assert diff_01_02 > 0, "Углы должны меняться при roll=0.2"
    print("  ✅ test_ik_roll_varies_smoothly")


if __name__ == "__main__":
    print("=" * 60)
    print("Python IK с roll тесты")
    print("=" * 60)

    tests = [
        test_ik_zero_roll_default_stance,
        test_ik_roll_45_degrees_affects_angles,
        test_ik_roll_45_angles_in_valid_range,
        test_ik_negative_roll_45,
        test_ik_left_right_symmetry_zero_roll,
        test_fk_ik_roundtrip_zero_roll,
        test_ik_small_orientation_angles,
        test_ik_roll_varies_smoothly,
    ]

    passed = 0
    failed = 0
    for t in tests:
        try:
            t()
            passed += 1
        except Exception as e:
            print(f"  ❌ {t.__name__}: {e}")
            failed += 1

    print("-" * 60)
    print(f"Python IK с roll: {passed}/{passed + failed} ✅, {failed} ❌")
    print("=" * 60)
    sys.exit(0 if failed == 0 else 1)

File: walking_robot_sim/src/quadropted_controller_cpp/test/test_ik_with_roll.cpp
#include <gtest/gtest.h>

#include "quadropted_controller_cpp/kinematics/forward_kinematics.hpp"
#include "quadropted_controller_cpp/kinematics/inverse_kinematics.hpp"

// ════════════════════════════════════════════════════════════
// IK с roll тесты — проверка влияния крена на углы суставов
// ════════════════════════════════════════════════════════════

// Тест 1: IK с roll=0 для default_stance — базовый случай
TEST(IKWithRoll, zero_roll_default_stance) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    auto angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0);
    ASSERT_EQ(angles.size(), 12u);

    // Углы должны быть в реалистичном диапазоне для стойки
    // Python: [0.000, 0.862, -1.883] для первой ноги
    EXPECT_NEAR(angles[0], 0.0, 0.01);
    EXPECT_NEAR(angles[1], 0.862, 0.1);
    EXPECT_NEAR(angles[2], -1.883, 0.1);
}

// Тест 2: IK с roll=π/4 — проверка влияния 45° крена
TEST(IKWithRoll, roll_45_degrees_affects_joint_angles) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    auto angles_no_roll = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0);
    auto angles_roll_45 = ik.inverse_kinematics(stance, 0, 0, 0.25, M_PI / 4, 0, 0);

    // Углы должны отличаться при наличии крена
    bool different = false;
    for (int i = 0; i < 12; ++i) {
        if (std::abs(angles_no_roll[i] - angles_roll_45[i]) > 1e-6) {
            different = true;
            break;
        }
    }
    EXPECT_TRUE(different) << "IK с roll=π/4 должен давать другие углы чем roll=0";
}

// Тест 3: IK с roll=0, pitch=0, yaw=0 — идентичность Python
TEST(IKWithRoll, matches_python_zero_orientation) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    auto angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0);

    // Python значения для REST стойки
    EXPECT_NEAR(angles[0], 0.0, 0.01);    // hip FR
    EXPECT_NEAR(angles[1], 0.862, 0.1);   // thigh FR
    EXPECT_NEAR(angles[2], -1.883, 0.1);  // calf FR

    EXPECT_NEAR(angles[3], 0.0, 0.01);    // hip FL
    EXPECT_NEAR(angles[4], 0.862, 0.1);   // thigh FL
    EXPECT_NEAR(angles[5], -1.883, 0.1);  // calf FL
}

// Тест 4: IK roundtrip: FK → IK → те же углы (с roll=0)
TEST(IKWithRoll, fk_ik_roundtrip_zero_roll) {
    double body[] = {0.3762, 0.0935};
    double legs[] = {0.0, 0.0955, 0.213, 0.213};

    quadropted::ForwardKinematics fk(body[0], body[1], legs[0], legs[1], legs[2], legs[3]);
    quadropted::InverseKinematics ik(body[0], body[1], legs[0], legs[1], legs[2], legs[3]);

    // Исходные углы
    std::vector<double> original_angles = {0, 0.3, -0.6, 0, 0.3, -0.6, 0, 0.3, -0.6, 0, 0.3, -0.6};

    // FK: углы → позиции ног
    auto foot_pos = fk.forward_kinematics_all_legs(original_angles);
    Eigen::MatrixXd leg_positions(3, 4);
    for (int leg = 0; leg < 4; ++leg) {
        for (int dim = 0; dim < 3; ++dim) {
            leg_positions(dim, leg) = foot_pos[leg](dim);
        }
    }

    // IK: позиции ног → углы
    auto recovered_angles = ik.inverse_kinematics(leg_positions, 0, 0, 0.25, 0, 0, 0);

    // Проверяем что углы восстановлены
    for (int i = 0; i < 12; ++i) {
        EXPECT_NEAR(recovered_angles[i], original_angles[i], 0.01) << "Angle " << i << " mismatch";
    }
}

// Тест 5: IK с roll=π/4 — углы в допустимом диапазоне
TEST(IKWithRoll, roll_45_angles_in_valid_range) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    auto angles = ik.inverse_kinematics(stance, 0, 0, 0.25, M_PI / 4, 0, 0);

    // Все углы должны быть в пределах ±π
    for (int i = 0; i < 12; ++i) {
        EXPECT_LT(std::abs(angles[i]), M_PI + 0.5) << "Angle " << i << " out of range: " << angles[i];
    }
}

// Тест 6: Симметрия углов — левые и правые ноги зеркальны при roll=0
TEST(IKWithRoll, left_right_symmetry_zero_roll) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    auto angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0, 0, 0);

    // FR (0,1,2) и FL (3,4,5) должны быть симметричны по hip
    // Hip углы: FR и FL должны быть одинаковы по модулю, противоположны по знаку
    // Но для symmetric stance они должны быть одинаковы
    EXPECT_NEAR(angles[0], angles[3], 0.01);  // hip FR vs hip FL
    EXPECT_NEAR(angles[1], angles[4], 0.01);  // thigh FR vs thigh FL
    EXPECT_NEAR(angles[2], angles[5], 0.01);  // calf FR vs calf FL
}

// Тест 7: IK с roll=-π/4 — отрицательный крен
TEST(IKWithRoll, negative_roll_45) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    auto angles_pos = ik.inverse_kinematics(stance, 0, 0, 0.25, M_PI / 4, 0, 0);
    auto angles_neg = ik.inverse_kinematics(stance, 0, 0, 0.25, -M_PI / 4, 0, 0);

    // Углы при +roll и -roll должны быть разными
    bool different = false;
    for (int i = 0; i < 12; ++i) {
        if (std::abs(angles_pos[i] - angles_neg[i]) > 1e-6) {
            different = true;
            break;
        }
    }
    EXPECT_TRUE(different) << "IK с roll=+π/4 и roll=-π/4 должны давать разные углы";
}

// Тест 8: IK с roll=0.1, pitch=-0.05, yaw=0.02 — совпадение с Python
TEST(IKWithRoll, matches_python_small_angles) {
    quadropted::InverseKinematics ik(0.3762, 0.0935, 0.0, 0.0955, 0.213, 0.213);

    double dx = 0.3762 * 0.5 + 0.02;
    double dy = 0.0935 * 0.5 + 0.0955;
    Eigen::MatrixXd stance(3, 4);
    stance << dx, dx, -dx, -dx, -dy, dy, -dy, dy, 0, 0, 0, 0;

    // Python: rotxyz(0.1, -0.05, 0.02) → конкретная матрица
    auto angles = ik.inverse_kinematics(stance, 0, 0, 0.25, 0.1, -0.05, 0.02);

    // Проверка что углы в реалистичном диапазоне
    for (int i = 0; i < 12; ++i) {
        EXPECT_LT(std::abs(angles[i]), M_PI + 0.5) << "Angle " << i << " out of range";
    }
}
